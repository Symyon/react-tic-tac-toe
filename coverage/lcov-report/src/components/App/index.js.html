<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/components/App/index.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="..\..\..\prettify.css" />
    <link rel="stylesheet" href="..\..\..\base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(..\..\..\sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="..\..\..\index.html">All files</a> / <a href="index.html">src/components/App</a> index.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">96.77% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>60/62</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">86.49% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>32/37</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>8/8</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>53/53</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">4x</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">21x</span>
<span class="cline-any cline-yes">13x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">10x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">6x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">9x</span>
<span class="cline-any cline-yes">7x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">3x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">16x</span>
<span class="cline-any cline-yes">16x</span>
<span class="cline-any cline-yes">144x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">16x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">15x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">24x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import React, { Component } from 'react';
import Board from '../Board';
import Score from '../Score';
import Button from '../Button';
import nought from './img/o.png';
import cross from './img/x.png';
import { EMPTY, NOUGHT, CROSS, DRAW } from '../../constants';
&nbsp;
import './index.css';
&nbsp;
class App extends Component {
&nbsp;
  constructor(props) {
    super(props);
&nbsp;
    this.state = {
      game: [],
      moves: [],
      crossActive: true,
      xScore: 0,
      oScore: 0,
      winner: EMPTY,
      winningCells: ''
    }
&nbsp;
    this.undo = this.undo.bind(this);
    this.restart = this.restart.bind(this);
    this.makeMove = this.makeMove.bind(this);
  }
&nbsp;
  makeMove(index) {
    const { game, moves, crossActive, winner } = this.state;
    if (winner !== EMPTY || game[index] !== EMPTY) return;
&nbsp;
    moves.push(index);
    game[index] = crossActive ? CROSS : <span class="branch-1 cbranch-no" title="branch not covered" >NOUGHT;</span>
    this.setState({ game, moves, crossActive: !crossActive })
&nbsp;
    <span class="missing-if-branch" title="else path not taken" >E</span>if (moves.length &gt;= 3) {
      const { matchMe, winningCells } = this.checkGameStatus(this.state.game, this.state.moves);
      this.markWinner(matchMe, winningCells);
    }
  }
&nbsp;
  checkGameStatus(game, moves) {
&nbsp;
    let winningCells;
    //check rows   
    for (let i = 0; i &lt; 3; i++) {
      const matchMe = game[i * 3];
      winningCells = '';
      let j;
      for (j = 0; j &lt; 3; j++) {
        const index = i * 3 + j;
        if (game[index] === EMPTY || game[index] !== matchMe) break;
        winningCells += index;
      }
      if (j === 3) return { matchMe, winningCells };
    }
&nbsp;
    //check columns
    for (let i = 0; i &lt; 3; i++) {
      const matchMe = game[i];
      winningCells = '';
      let j;
      for (j = 0; j &lt; 3; j++) {
        const index = i + j * 3;
        if (game[index] === EMPTY || game[index] !== matchMe) break;
        winningCells += index;
      }
      <span class="missing-if-branch" title="if path not taken" >I</span>if (j === 3) <span class="cstat-no" title="statement not covered" >return { matchMe, winningCells };</span>
    }
&nbsp;
    //check diagonals
    for (let i = 0; i &lt; 3; i += 2) {
      const matchMe = game[i * 3];
      winningCells = '';
      let j;
      for (j = 0; j &lt; 3; j++) {
        const index = Math.abs(i - j) * 3 + j;
        if (game[index] === EMPTY || game[index] !== matchMe) break;
        winningCells += index;
      }
      if (j === 3) return { matchMe, winningCells };
    }
&nbsp;
    return { matchMe: moves.length === 9 ? DRAW : <span class="branch-1 cbranch-no" title="branch not covered" >EMPTY,</span> winningCells: '' };
  }
&nbsp;
  markWinner(matchMe, winningCells) {
    let { xScore, oScore } = this.state;
    if (matchMe === CROSS) xScore++;
    else <span class="missing-if-branch" title="if path not taken" >I</span>if (matchMe === NOUGHT) <span class="cstat-no" title="statement not covered" >oScore++;</span>
&nbsp;
    this.setState({ winner: matchMe, xScore, oScore, winningCells })
  }
&nbsp;
  undo() {
    const { game, moves, crossActive } = this.state;
&nbsp;
    game[moves.pop()] = EMPTY;
    this.setState({ game, moves, crossActive: !crossActive })
  }
&nbsp;
  restart() {
    let arr = new Array(9);
    for (let i = 0; i &lt; 9; i++) {
      arr[i] = EMPTY;
    }
&nbsp;
    this.setState({ game: arr, moves: [], crossActive: true, winner: EMPTY, winningCells: '' })
  }
&nbsp;
&nbsp;
  componentWillMount() {
    this.restart();
  }
&nbsp;
  render() {
&nbsp;
    const { moves, winner } = this.state;
    const isUndo = moves.length !== 0 &amp;&amp; moves.length !== 9 &amp;&amp; winner === EMPTY;
&nbsp;
    return (
      &lt;div className="App"&gt;
        &lt;div&gt;
          &lt;div className="ScoreContainer"&gt;
            &lt;Score icon={cross} value={this.state.xScore} active={this.state.crossActive} /&gt;
            &lt;Score icon={nought} value={this.state.oScore} active={!this.state.crossActive} /&gt;
          &lt;/div&gt;
          &lt;Board game={this.state.game} onClick={this.makeMove} winningCells={this.state.winningCells} winner={this.state.winner} /&gt;
          &lt;div className="ButtonsContainer"&gt;
            &lt;Button text='Undo' onClick={isUndo ? this.undo : null} /&gt;
            &lt;Button text='Restart' onClick={this.restart} /&gt;
          &lt;/div&gt;
        &lt;/div&gt;
      &lt;/div&gt;
    );
  }
}
&nbsp;
export default App;
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 27 2019 16:27:58 GMT-0500 (Eastern Standard Time)
</div>
</div>
<script src="..\..\..\prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="..\..\..\sorter.js"></script>
</body>
</html>
