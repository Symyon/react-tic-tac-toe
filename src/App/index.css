.App {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  background-color: blue
}

.ScoreContainer {
  display: flex;
  flex: 1;
  flex-direction: row;
  width: 100%
}

.ActivePlayer {

}

.ButtonsContainer {
  display: flex;
  flex: 1;
  flex-direction: row;
}