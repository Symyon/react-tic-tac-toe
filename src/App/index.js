import React, { Component } from 'react';
import Board from '../Board';
import Score from '../Score';
import Button from '../Button';

import './index.css';

class App extends Component {

  constructor(props) {
    super(props);

    this.undo = this.undo.bind(this);
    this.restart = this.restart.bind(this);
  }

  undo () {

  }

  restart () {

  }

  render() {
    return (
      <div className="App" style={{ display: 'flex', flexDirection: 'column', justifyContent: 'center', alignItems: 'center', backgroundColor: 'blue' }}>
        <div>
          <div style={{ display: 'flex', flex: 1, flexDirection: 'row', width: '100%' }}>
            <Score text = 'X score'/>
            <Score text = 'O score'/>
          </div>
          <Board />
          <div style={{ display: 'flex', flex: 1, flexDirection: 'row' }}>
            <Button text ='Undo' onClick = {this.undo}/>
            <Button text='Restart' onClick = {this.restart}/>
          </div>
        </div>
      </div>
    );
  }
}

export default App;
