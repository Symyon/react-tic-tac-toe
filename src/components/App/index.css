.App {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  background-color: blue;
  background-image: url("img/wall.jpg");
  background-size: 100%;
}

.ScoreContainer {
  display: flex;
  flex: 1;
  flex-direction: row;
  width: 100%
}

.ActivePlayer {

}

.ButtonsContainer {
  display: flex;
  flex: 1;
  flex-direction: row;
}