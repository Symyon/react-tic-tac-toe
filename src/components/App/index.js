import React, { Component } from 'react';
import Board from '../Board';
import Score from '../Score';
import Button from '../Button';
import nought from './img/o.png';
import cross from './img/x.png';
import {EMPTY, NOUGHT, CROSS, DRAW} from '../../constants';

import './index.css';

class App extends Component {

  constructor(props) {
    super(props);

    this.state = {
      game: [],
      moves: [],
      crossActive: true,
      xScore: 0,
      oScore: 0,
      winner: EMPTY
    }

    this.undo = this.undo.bind(this);
    this.restart = this.restart.bind(this);
    this.makeMove = this.makeMove.bind(this);
  }

  makeMove(index) {
    const { game, moves, crossActive, winner } = this.state;
    if (winner !== EMPTY || game[index] !== EMPTY) return;

    moves.push(index);
    game[index] = crossActive ? CROSS : NOUGHT;
    this.setState({ game, moves, crossActive: !crossActive })
    if (moves.length >= 3) this.checkGameStatus();
  }

  checkGameStatus() {
    const { game, moves } = this.state;

    //check rows   
    for (let i = 0; i < 3; i++) {
      const matchMe = game[i * 3];
      let j;
      for (j = 0; j < 3; j++) {
        const index = i * 3 + j;
        if (game[index] === EMPTY || game[index] !== matchMe) break;
      }
      if (j === 3) { this.markWinner(matchMe); return; }
    }

    //check columns
    for (let i = 0; i < 3; i++) {
      const matchMe = game[i];
      let j;
      for (j = 0; j < 3; j++) {
        const index = i + j * 3;
        if (game[index] === EMPTY || game[index] !== matchMe) break;
      }
      if (j === 3) { this.markWinner(matchMe); return; }
    }
   
    //check diagonals
    for (let i = 0; i < 3; i+=2) {
      const matchMe = game[i * 3];
      let j;
      for(j = 0; j < 3; j++) {
        const index = Math.abs(i - j) * 3 + j;
        if (game[index] === EMPTY || game[index] !== matchMe) break;
      }
      if (j === 3) { this.markWinner(matchMe); return; }
    }
    
    if(moves.length === 9) this.markWinner(DRAW);
  }

  markWinner(matchMe) {
    let { xScore, oScore } = this.state;
    if (matchMe === CROSS) xScore++;
    else if (matchMe === NOUGHT) oScore++;

    this.setState({ winner: matchMe, xScore, oScore })
  }

  undo() {
    const { game, moves, crossActive } = this.state;

    game[moves.pop()] = EMPTY;
    this.setState({ game, moves, crossActive: !crossActive })
  }

  restart() {
    let arr = new Array(9);
    for (let i = 0; i < 9; i++) {
      arr[i] = EMPTY;
    }

    this.setState({ game: arr, moves: [], crossActive: true, winner: EMPTY })
  }


  componentWillMount() {
    this.restart();
  }

  render() {

    const { moves, winner } = this.state;
    const isUndo = moves.length !== 0 && moves.length !== 9 && winner === EMPTY;

    return (
      <div className="App">
        <div>
          <div className="ScoreContainer">
            <Score icon={cross} text={`Score: ${this.state.xScore}`} active={this.state.crossActive} />
            <Score icon={nought} text={`Score: ${this.state.oScore}`} active={!this.state.crossActive}/>
          </div>
          <Board game={this.state.game} onClick={this.makeMove} />
          <div className="ButtonsContainer">
            <Button text='Undo' onClick={this.undo} active = {isUndo} />
            <Button text='Restart' onClick={this.restart} active = {true}/>
          </div>
          {this.state.winner !== EMPTY && <span>Winner is {this.state.winner}</span>}
        </div>
      </div>
    );
  }
}

export default App;
