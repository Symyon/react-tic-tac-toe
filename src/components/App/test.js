import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, render } from 'enzyme';
import App from '.';

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const component = shallow(<App />);
    expect(component).toMatchSnapshot();
  });

  it('renders 2 Score components', () => {
    const component = render(<App />);
    expect(component.find('div.Score').length).toEqual(2);
  })

  it('renders 1 Board component', () => {
    const component = render(<App />);
    expect(component.find('div.Board').length).toEqual(1);
  })
});
