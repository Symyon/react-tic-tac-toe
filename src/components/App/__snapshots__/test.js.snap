// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="App"
>
  <div>
    <div
      className="ScoreContainer"
    >
      <Score
        active={true}
        icon="x.png"
        value={0}
      />
      <Score
        active={false}
        icon="o.png"
        value={0}
      />
    </div>
    <Board
      game={
        Array [
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
          -1,
        ]
      }
      onClick={[Function]}
      winner={-1}
      winningCells=""
    />
    <div
      className="ButtonsContainer"
    >
      <Button
        onClick={null}
        text="Undo"
      />
      <Button
        onClick={[Function]}
        text="Restart"
      />
    </div>
  </div>
</div>
`;
